File: Java/src/main/java/GruesomeCave.java
import java.util.*;

public class GruesomeCave {
    public static void main(String[] args) {
        Scanner scn = new Scanner(System.in);

        int L = scn.nextInt();
        int W = scn.nextInt();
        scn.nextLine();
        char[][] chars = new char[L][W];

        for (int l = 0; l < L; l++) chars[l] = scn.nextLine().toCharArray();

        System.out.println(calculateGrueRisk(chars));
    }

    /**
     * Using djikstra to find minimum path, and calculate the risk
     * @param maze
     * @return
     */
    public static double calculateGrueRisk(char[][] maze) {
        var risksTable = getRiskTable(maze);
        var totalRisk = getCaveInfo(risksTable);
        int[][] distTable = new int[maze.length][maze[0].length];

        for (int i = 0; i < distTable.length; i++)
            for (int j = 0; j < distTable[0].length; j++)
                distTable[i][j] = Integer.MAX_VALUE;

        distTable[totalRisk.start.y][totalRisk.start.x] = 0;
        PriorityQueue<Pos> q = new PriorityQueue<>();
        totalRisk.start.value = 0;
        q.add(totalRisk.start);

        while (!q.isEmpty()) {
            var cx = q.peek().x;
            var cy = q.peek().y;
            q.poll();

            for (var p : getNeighbohurs(cx, cy, maze))
                if (risksTable[p.y][p.x] > -1)
                    if (distTable[p.y][p.x] > distTable[cy][cx] + risksTable[cy][cx]) {
                        distTable[p.y][p.x] = distTable[cy][cx] + risksTable[cy][cx];
                        p.value = distTable[p.y][p.x];
                        q.add(p);
                    }
        }


        if (totalRisk.maxRisk == 0) return 0;
        return (double) distTable[totalRisk.stop.y][totalRisk.stop.x] / (double) totalRisk.maxRisk;
    }

    public static CaveInfo getCaveInfo(int[][] risks) {
        var res = new CaveInfo();
        for (int i = 0; i < risks.length; i++)
            for (int j = 0; j < risks[0].length; j++)
                if (risks[i][j] > 0)
                    res.maxRisk += risks[i][j];
                else if (risks[i][j] == 0)
                    if (res.start == null) res.start = new Pos(j, i);
                    else res.stop = new Pos(j, i);
        return res;
    }

    public static int[][] getRiskTable(char[][] maze) {
        int[][] risks = new int[maze.length][maze[0].length];
        for (int i = 0; i < maze.length; i++) {
            for (int j = 0; j < maze[0].length; j++) {
                switch (maze[i][j]) {
                    case '#':
                        risks[i][j] = -1;
                        break;
                    case 'E':
                    case 'D':
                        risks[i][j] = 0;
                        break;
                    case ' ':
                        for (Pos p : getNeighbohurs(i, j, maze)) if (maze[p.y][p.x] == ' ') risks[p.y][p.x]++;
                        break;
                    default:
                        throw new IllegalStateException("Unexpected value: " + maze[i][j]);
                }
            }
        }
        return risks;
    }

    public static List<Pos> getNeighbohurs(int cx, int cy, char[][] maze) {
        ArrayList<Pos> valids = new ArrayList<>();
        int[] x = {0, +1, 0, -1};
        int[] y = {+1, 0, -1, 0};
        for (int i = 0; i < 4; i++)
            try {
                if (maze[cy - y[i]][cx - x[i]] != '#')
                    valids.add(new Pos(cx - x[i], cy - y[i]));
            } catch (IndexOutOfBoundsException ignore) {
            }
        return valids;
    }
}

class Pos implements Comparable<Pos> {
    int x, y;
    long value;

    public Pos(int x, int y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public int compareTo(Pos o) {
        return Long.compare(this.value, o.value);
    }
}

class CaveInfo {
    Pos start, stop;
    long maxRisk;
}

File: C++/src/GruesomeCave.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <set>
#include <map>
#include <queue>
#include <unordered_set>
#include <string>
#include <climits>

using namespace std;

class Pos {
public:
	int x, y;
	long value;

	bool operator<(const Pos& a) const {
		return (value > a.value);
	}

};

class CaveInfo {
public:
	Pos start, stop;
	long maxRisk;
	bool startSet;
};


static CaveInfo* getCaveInfo(vector<vector<int>>* risks) {
	CaveInfo* res = new CaveInfo();
	Pos p;
	for (int i = 0; i < risks->size(); i++)
		for (int j = 0; j < risks->at(i).size(); j++)
			if (risks->at(i).at(j) > 0)
				res->maxRisk += risks->at(i).at(j);
			else if (risks->at(i).at(j) == 0) {
				p.x = j;
				p.y = i;
				if (!res->startSet) {
					res->start = p;
					res->startSet = true;
				}
				else res->stop = p;
			}
	return res;
}

static bool isInside(int* x, int* y, vector<vector<char>>* maze) {
	return (*x >= 0 && *y >= 0 && *x < maze->size() && *y < maze->at(*x).size());
}

static vector<Pos>* getNeighbohurs(int* cx, int* cy, vector<vector<char>>* maze) {
	vector<Pos>* valids = new vector<Pos>();
	Pos newPos;
	const int x[] = { 0, +1, 0, -1 };
	const int y[] = { +1, 0, -1, 0 };

	for (int i = 0; i < 4; i++) {
		newPos.x = *cx - x[i];
		newPos.y = *cy - y[i];

		if (!isInside(&newPos.x, &newPos.y, maze)) continue;

		if (maze->at(newPos.y).at(newPos.x) != '#')
			valids->push_back(newPos);
	}
	return valids;
}

static vector<vector<int>>* getRiskTable(vector<vector<char>>* maze) {
	vector<vector<int>>* risks = new vector<vector<int>>();

	for (int i = 0; i < maze->size(); i++) {
		vector<int> r;
		for (int j = 0; j < maze->at(i).size(); j++)
			r.push_back(0);
		risks->push_back(r);
	}

	for (int i = 0; i < maze->size(); i++) {
		for (int j = 0; j < maze->at(i).size(); j++) {
			switch (maze->at(i).at(j)) {
			case '#':
				risks->at(i).at(j) = -1;
				break;
			case 'E':
			case 'D':
				risks->at(i).at(j) = 0;
				break;
			case ' ':
				for (auto p : *getNeighbohurs(&j, &i, maze)) {
					if (maze->at(p.y).at(p.x) == ' ')
						risks->at(p.y).at(p.x)++;
				}
				break;
			default:
				throw new exception;
			}
		}
	}
	return risks;
}

/**
 * Using djikstra to find minimum path, and calculate the risk
 * @param maze
 * @return
 */
static double calculateGrueRisk(vector<vector<char>>* maze) {
	vector<vector<int>>* risksTable = getRiskTable(maze);
	CaveInfo* totalRisk = getCaveInfo(risksTable);

	vector<vector<int>> distTable;

	for (int i = 0; i < maze->size(); i++) {
		vector<int> r;
		for (int j = 0; j < maze->at(i).size(); j++)
			r.push_back(INT_MAX);
		distTable.push_back(r);
	}

	distTable.at(totalRisk->start.y).at(totalRisk->start.x) = 0;
	priority_queue<Pos> q;
	totalRisk->start.value = 0;
	q.push(totalRisk->start);

	while (!q.empty()) {
		int cx = q.top().x;
		int cy = q.top().y;
		q.pop();

		for (auto& p : *getNeighbohurs(&cx, &cy, maze))
			if (risksTable->at(p.y).at(p.x) > -1)
				if (distTable.at(p.y).at(p.x) > distTable.at(cy).at(cx) + risksTable->at(p.y).at(p.x)) {
					distTable.at(p.y).at(p.x) = distTable.at(cy).at(cx) + risksTable->at(p.y).at(p.x);
					p.value = distTable.at(p.y).at(p.x);
					q.push(p);
				}
	}

	if (totalRisk->maxRisk == 0) return 0;
	return (double)distTable.at(totalRisk->stop.y).at(totalRisk->stop.x) / (double)totalRisk->maxRisk;
}

int main() {
	std::ios::sync_with_stdio(false);
	std::cin.tie(NULL);

	int L, W;
	string tmp;
	cin >> L >> W;
	vector<vector<char>> chars;
	getline(cin, tmp);

	for (int l = 0; l < L; l++) {
		getline(cin, tmp);
		vector<char> line;
		for (int w = 0; w < W; w++) line.push_back(tmp.at(w));
		chars.push_back(line);
	}

	cout << calculateGrueRisk(&chars) << endl;
}